### ISSUE TITLE
Suspected missing memory barrier on Aarch64

### ISSUE THREAD

**Issue by eboasson** [eclipse-iceoryx/iceoryx#2006]
## Required information

**Operating system:**
macOS 13.5 and Ubuntu 22.04 LTS

**Compiler version:**
Apple clang 14.0.3, gcc 11.4.0

**CPU:**
Apple M1

**Eclipse iceoryx version:**
`v2.0.3`

**Observed result or behaviour:**
Incorrect error messages like:
```
2023-08-19 18:15:44.232 [ Error ]: MemoryManager: unable to acquire a chunk with a chunk-payload size of 67776The following mempools are available:  MemPool [ ChunkSize = 168, ChunkPayloadSize = 128, ChunkCount = 10000 ]  MemPool [ ChunkSize = 1064, ChunkPayloadSize = 1024, ChunkCount = 5000 ]  MemPool [ ChunkSize = 16424, ChunkPayloadSize = 16384, ChunkCount = 1000 ]  MemPool [ ChunkSize = 131112, ChunkPayloadSize = 131072, ChunkCount = 200 ]  MemPool [ ChunkSize = 524328, ChunkPayloadSize = 524288, ChunkCount = 50 ]  MemPool [ ChunkSize = 1048616, ChunkPayloadSize = 1048576, ChunkCount = 30 ]  MemPool [ ChunkSize = 4194344, ChunkPayloadSize = 4194304, ChunkCount = 10 ]
2023-08-19 18:15:44.232 [Warning]: ICEORYX error! MEPOO__MEMPOOL_GETCHUNK_POOL_IS_RUNNING_OUT_OF_CHUNKS
=== [ICEORYX] failure getting loan: AllocationError::RUNNING_OUT_OF_CHUNKS
```
I have verified the sequence of chunks being requested/taken from a subscriber and publications/releases and it checks out. There is a mutex around the operations to prevent multi-threaded access to a single Iceoryx object.

**Expected result or behaviour:**
Not running out of chunks.

**Conditions where it occurred / Performed steps:**
Using: https://github.com/eboasson/cyclonedds/commit/51e1505f71fcc256d40a91dc620a31e4c899b0cf (I created a branch to keep the commit alive, so the link should be reasonably stable) and acknowledging there are some hacks in there that I am not proud of (but I wanted to try out some things 🙂), running:
```
CYCLONEDDS_URI="$CYCLONEDDS_URI,<Share><Enable>true</>" bin/Debug/ddsperf -L -TS32k -k 1 pub sub
```
on an M1-based computer, either natively on macOS or using UTM + Ubuntu should trigger it.

An addition and important piece of information: if I build Iceoryx and Cyclone for macOS x64 instead of macOS aarch64, and thereby triggering Rosetta-based machine code translation and switching the CPU to "total store order" mode (matching an x64 CPU unlike the weaker ordering native to ARM) the problem disappears.

Which makes me suspect there is a memory barrier missing somewhere in the Iceoryx code.

Regarding the checking of operations: reverting the HEAD, redirecting the output from the command to a file, say `log.txt` and doing:
```
perl -lane 'next unless /^([+-])([WR])/;$g=($2 eq "R"?2:1);$o=($1 eq "+"?$g:0);$s+=($o>0?1:-1);$ms=($s>$ms?$s:$ms);if($m{$F[2]}!=$g-$o){print "max=$ms s=$s\n";while(($k,$v) = each %m){print "$k $v" if $v};die;};$m{$F[2]}=$o;' log.txt
```
should show at a maximum of 3 outstanding memory chunks since start up at the point in time that Iceoryx reports running out of chunks.

**Comment by elBoberido** [comment]
@eboasson is it possible to check if the issue still persists with the latest master? In theory the C-binding should (again) be source compatible with v2.0.3.

Would it be possible to run the introspection client in parallel and check how the chunk count changes over time. https://github.com/elBoberido/iceray would show a nice graph but I think I have to create a new release in order to work with v2.0.3. Currently it depends on v2.0.2

**Comment by eboasson** [comment]
Hi @elBoberido, I have a confession to make ...

I let the nemesis of the software engineer get the better of me. I made an assumption and forgot to check it ... Mea culpa.

So that branch I linked to has a completely rewritten integration of Iceoryx, where the interface on the Cyclone side has been generalised a bit so that it can load the Iceoryx integration as a plugin. The integrated plugin is written by someone else (in C++, so the C binding is not needed for this variant of Cyclone) but it turns out I assumed incorrectly that it would map the QoS settings correctly. Instead it uses the default pub/sub options, which means the subscriber has `MAX_QUEUE_CAPACITY` even though my test only requires a keep-last-1 policy.

A quick look did not tell me what `MAX_QUEUE_CAPACITY` is for a largish chunk size with the default RouDI settings, but I find it entirely plausible that it would be larger than the number of chunks actually available.

The rest would then be scheduling and the ARM vs x64_86 bit would just be thing my old nemesis would use to hide a subtle difference in scheduling and make it appear "hardware" dependent ...

I just added setting the publisher and subscriber options and it ran for several minutes without any problem, whereas before it would fail typically within 30s.

Apologies for the scare!
